topdown/example_test.go

```diff
@@ -147,3 +150,80 @@ func ExampleQuery() {
 	// err1: <nil>
 
 }
+
+func ExampleRegisterFunctionalBuiltin1() {
+
+	// Rego includes a number of built-in functions ("built-ins") for performing
+	// standard operations like string manipulation, regular expression
+	// matching, and computing aggregates.
+	//
+	// This test shows how to add a new built-in to Rego and OPA.
+
+	// Initialize context for the example. Normally the caller would obtain the
+	// context from an input parameter or instantiate their own.
+	ctx := context.Background()
+
+	// The ast package contains a registry that enumerates the built-ins
+	// included in Rego. When adding a new built-in, you must update the
+	// registry to include your built-in. Otherwise, the compiler will complain
+	// when it encounters your built-in.
+	builtin := &ast.Builtin{
+		Name:      ast.Var("selective_upper"),
+		NumArgs:   2,
+		TargetPos: []int{1},
+	}
+
+	ast.RegisterBuiltin(builtin)
+
+	// This is the implementation of the built-in that will be called during
+	// query evaluation.
+	builtinImpl := func(a ast.Value) (ast.Value, error) {
+
+		str, err := builtins.StringOperand(a, 1)
+
+		if err != nil {
+			return nil, err
+		}
+
+		if str.Equal(ast.String("magic")) {
+			// topdown.BuiltinEmpty indicates to the evaluation engine that the
+			// expression is false/not defined.
+			return nil, topdown.BuiltinEmpty{}
+		}
+
+		return ast.String(strings.ToUpper(string(str))), nil
+	}
+
+	// See documentation for registering functions that take different numbers
+	// of arguments.
+	topdown.RegisterFunctionalBuiltin1(builtin.Name, builtinImpl)
```

Comment:
I like how simple this is.  The comments earlier about having a single registration function would be nice, but not necessary.